urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/driver/examples/INDIElTiempoDriver.java

```diff
@@ -0,0 +1,451 @@
+      }
+    }
+
+    byte[] imageBytes;
+
+    try {
+      imageBytes = readBinaryFile(image);
+    } catch (IOException e) {
+      e.printStackTrace();
+
+      return false;
+    }
+
+    // Define
+    INDIBLOBProperty pim = (INDIBLOBProperty) getProperty(imagePrefix + "_SATELLITE_IMAGE");
+    INDIBLOBElement eim = (INDIBLOBElement) pim.getElement(imagePrefix + "_SATELLITE_IMAGE");
+
+    if (Arrays.equals(imageBytes, eim.getValue().getBLOBData())) {
+      return false;  // The same image as the one in the property
+    }
+
+    eim.setValue(new INDIBLOBValue(imageBytes, "jpg"));
+
+    int pos1 = fileName.lastIndexOf("-");
+
+    String name = fileName.substring(pos1, pos1 + 5) + "/" + fileName.substring(pos1 + 5, pos1 + 7) + "/" + fileName.substring(pos1 + 7, pos1 + 9) + " " + fileName.substring(pos1 + 9, pos1 + 11) + ":" + fileName.substring(pos1 + 11, pos1 + 13);
+
+    INDITextProperty pn = (INDITextProperty) getProperty(imagePrefix + "_IMAGE_NAME");
+    INDITextElement en = (INDITextElement) pn.getElement(imagePrefix + "_IMAGE_NAME");
+
+    en.setValue(imagePrefix + " Satellite " + name + " UTC");
+
+    return true;
+  }
+
+  /**
+   * Reads a text file and returns its contents as a String.
+   *
+   * @param file The file to read
+   * @return The contents of the file
+   * @throws IOException
+   */
+  private String readFile(File file) throws IOException {
+    BufferedReader reader = new BufferedReader(new FileReader(file));
+    String line;
+    StringBuilder stringBuilder = new StringBuilder();
+    String ls = System.getProperty("line.separator");
+
+    while ((line = reader.readLine()) != null) {
+      stringBuilder.append(line);
+      stringBuilder.append(ls);
+    }
+
+    return stringBuilder.toString();
+  }
+
+  /**
+   * Reads a binary file and returns its contents as a byte[]
+   *
+   * @param file The file to be read
+   * @return The contents of the file
+   * @throws IOException
+   */
+  private byte[] readBinaryFile(File file) throws IOException {
+    int fileSize = (int) file.length();
+    FileInputStream reader = new FileInputStream(file);
+
+    byte[] buffer = new byte[fileSize];
+
+    int totalRead = 0;
+
+    while (totalRead < fileSize) {
+      int readed = reader.read(buffer, totalRead, fileSize - totalRead);
+
+      if (readed == -1) {
+        return null; // Unexpected end of file 
+      }
+
+      totalRead += readed;
+    }
+
+    return buffer;
+  }
+
+  /**
+   * Downloads a
+   * <code>url</code> and saves its contents to
+   * <code>file</code>.
+   *
+   * @param url
+   * @param file
+   * @throws IOException
+   * @throws MalformedURLException
+   */
+  private void downloadAndSave(String url, File file) throws IOException, MalformedURLException {
+    int bufsize = 65536;
+    byte[] buffer = new byte[bufsize];
+
+    URL u = new URL(url);
+    InputStream is = u.openStream();  // throws an IOException
+    BufferedInputStream bis = new BufferedInputStream(is);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *bis*. The resource is referenced at lines 388, 394. The resource is closed at line 401. There are other execution paths that do not contain closure statements, for example, when FileOutputStream constructor throws an exception. Ensure *bis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).